src/test/directives/migration-directives_test.ts

```diff
@@ -0,0 +1,206 @@
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+/// <reference path="../../../node_modules/@types/mocha/index.d.ts" />
+/// <reference path="../../../node_modules/@types/chai/index.d.ts" />
+
+import {Directive, directive, DirectiveParameters, Part, PartType,} from '../../directive.js';
+import {render} from '../../lib/render.js';
+import {html} from '../../lit-html.js';
+import {stripExpressionMarkers} from '../test-utils/strip-markers.js';
+
+const assert = chai.assert;
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+/**
+ * adds the length of the parent /current node's tag name to the given value on
+ * update
+ */
+class AddParentLengthDirective extends Directive {
+  _targetEl: Element|null = null;
+
+  render(value: number) {
+    return value.toString();
+  }
+
+  update(part: Part, [value]: DirectiveParameters<this>) {
+    const hasPrevTarget = !!this._targetEl;
+
+    if (!hasPrevTarget) {
+      switch (part.type) {
+        case PartType.ATTRIBUTE:
+        case PartType.PROPERTY:
+          this._targetEl = part.element;
+          break;
+        case PartType.CHILD:
+          this._targetEl = part.parentNode as Element;
+          break;
+        default:
+          throw new Error(
+              `${part.name} is not supported for parent length directive`);
+      }
+    }
+
+    const parentNameLength = this._targetEl!.tagName.length;
+    return this.render(value + parentNameLength);
+  }
+}
+
+/**
+ * concats the parent / current element's tag name to the given value on update
+ */
+class AddParentNameDirective extends Directive {
+  _targetEl: Element|null = null;
+
+  render(value: string|number) {
+    return value.toString();
+  }
+
+  update(part: Part, [value]: DirectiveParameters<this>) {
+    const hasPrevTarget = !!this._targetEl;
+
+    if (!hasPrevTarget) {
+      switch (part.type) {
+        case PartType.ATTRIBUTE:
+        case PartType.BOOLEAN_ATTRIBUTE:
+        case PartType.PROPERTY:
+          this._targetEl = part.element;
+          break;
+        case PartType.CHILD:
+          this._targetEl = part.parentNode as Element;
+          break;
+        default:
+          throw new Error(
+              `${part.name} is not supported for parent length directive`);
+      }
+    }
+
+    const parentName = this._targetEl!.tagName.toLocaleLowerCase();
+    return this.render(`${value} ${parentName}`);
+  }
+}
+
+const addParentLength = directive(AddParentLengthDirective);
+const addParentName = directive(AddParentNameDirective);
+
+suite('migration directives', () => {
```

Comment:
Should add a test verifying that directive state is maintained across re-renders.